### ISSUE TITLE
[pipelining] Interleaved1F1B warmup head-of-line bubble from front-loaded RECV

### ISSUE THREAD

**Issue by HAOCHENYE** [pytorch/pytorch#188833]
## TL;DR

On the `_PipelineScheduleRuntime` path (which `ScheduleInterleaved1F1B` now uses), the
communication lowering places each `RECV_F` at the position where its **sender** is scheduled
by a dependency simulation, **not** at the timestep where the data is actually produced/needed.

https://github.com/pytorch/pytorch/blob/main/torch/distributed/pipelining/schedules.py#L1536

During warmup this front-loads a long-latency `RECV` ahead of independent, already-ready `SEND`s
on the same rank.

I found a related issue:  #172668. However, I don't think this is AMD-specific and it should be optimized by an option — the ordering of the communication ops is itself unreasonable.                                                   

## Reproduction

A complete, self-contained repro: a stack of Linear+GELU blocks is split into `pp_size * vpp`
virtual stages and driven by `ScheduleInterleaved1F1B`. It prints rank 0's lowered communication
order at startup and captures one step with `torch.profiler`.

**8 gpus, pp 8, vpp 2**

```python
# toy_interleaved_hol.py  --  run with: torchrun --nproc_per_node=8 toy_interleaved_hol.py
import os

import torch
import torch.distributed as dist
import torch.nn as nn
import torch.nn.functional as F
from torch.distributed.pipelining import PipelineStage, ScheduleInterleaved1F1B

VPP = 2                 # virtual stages per rank
HIDDEN = 4096
LAYERS_PER_STAGE = 2
MB_BATCH = 1024
STEPS = 6
PROFILE_STEP = 4


class StageBlock(nn.Module):
    def __init__(self, hidden, layers):
        super().__init__()
        self.net = nn.Sequential(
            *[m for _ in range(layers) for m in (nn.Linear(hidden, hidden), nn.GELU())]
        )

    def forward(self, x):
        return self.net(x)


def main():
    rank = int(os.environ["RANK"])
    world = int(os.environ["WORLD_SIZE"])
    local_rank = int(os.environ["LOCAL_RANK"])

    pp = world                       # ep=1: world_size == pp_size
    mb = 2 * pp                      # multiple of number_of_rounds (=2) for interleaved-1F1B
    torch.cuda.set_device(local_rank)
    device = torch.device(f"cuda:{local_rank}")
    dist.init_process_group("nccl", device_id=device)

    total_stages = pp * VPP
    owned = list(range(rank, total_stages, pp))       # round-robin virtual stages
    last_rank = (total_stages - 1) % pp
    example = torch.randn(MB_BATCH, HIDDEN, device=device)

    stages = [
        PipelineStage(StageBlock(HIDDEN, LAYERS_PER_STAGE).to(device), s, total_stages, device,
                      input_args=example)
        for s in owned
    ]
    schedule = ScheduleInterleaved1F1B(stages, n_microbatches=mb, loss_fn=F.mse_loss)
    opt = torch.optim.SGD([p for s in stages for p in s.submod.parameters()], lr=1e-4)

    if rank == 0:
        print("# rank0 lowered comm order (first 12):", flush=True)
        for k, a in enumerate(schedule.pipeline_order_with_comms[0][:12]):
            print(f"  [{k:02d}] {a}", flush=True)

    x = torch.randn(mb * MB_BATCH, HIDDEN, device=device)     # chunked into mb microbatches
    y = torch.randn(mb * MB_BATCH, HIDDEN, device=device)

    def step_once():
        opt.zero_grad(set_to_none=True)
        losses = []
        if rank == 0:
            schedule.step(x)
        elif rank == last_rank:
            schedule.step(target=y, losses=losses)
        else:
            schedule.step()
        opt.step()

    prof = torch.profiler.profile(
        activities=[torch.profiler.ProfilerActivity.CPU, torch.profiler.ProfilerActivity.CUDA]
    )
    for step in range(STEPS):
        if step == PROFILE_STEP:
            dist.barrier()
            prof.start()
        step_once()
        torch.cuda.synchronize()
        if step == PROFILE_STEP:
            prof.stop()
            prof.export_chrome_trace(f"trace_rank{rank}.json")
    dist.destroy_process_group()


if __name__ == "__main__":
    main()
```

Run (`world_size == pp_size`, ep=1):

```bash
torchrun --nproc_per_node=8 toy_interleaved_hol.py
```

Then, I got the rank0 timeline like this:

<img width="1254" height="380" alt="Image" src="https://github.com/user-attachments/assets/cdc5f63e-1195-455e-b1e1-8a14ce92410d" />

and rank1 timeline:

<img width="1336" height="115" alt="Image" src="https://github.com/user-attachments/assets/61570f75-f20c-4315-9598-a997b57b6fb3" />

Obviously, the bubble of stage1's warmup phase is unreasonable.

At startup, rank 0 prints the lowered order the runtime executes:

```
# rank0 lowered comm order (first 12):
  [00] 0UNSHARD
  [01] 8UNSHARD
  [02] 0F0
  [03] 0SEND_F0
  [04] 8RECV_F0      <-- consumed only at t=8, but posted here, before 0SEND_F1
  [05] 0F1
  [06] 0SEND_F1      <-- ready (rank 1 needs it now) but queued AFTER 8RECV_F0
  [07] 8RECV_F1
  ...
```

`8RECV_F0` is queued at position [04], **ahead of** `0SEND_F1..0SEND_F6`, on rank 0's single serial
p2p stream. But `8RECV_F0` cannot complete until its match `7SEND_F0` arrives, and rank 7 only
issues `7SEND_F0` after it computes `7F0` at timestep 7 (the pipeline-fill diagonal) — i.e. ~7
warmup stages later. Because the stream is serial, `8RECV_F0`'s kernel holds it for that whole span,
so every `SEND` enqueued behind it — `0SEND_F1, 0SEND_F2, ...`, which rank 1 needs to start its
warmup forwards — **cannot execute until the recv completes**, even though those sends are ready
immediately. Worse, this early posting is pointless: the forward that actually consumes `8RECV_F0`,
`8F0`, does not run until timestep 8 (`0F0..0F7` occupy `t=0..7`, `8F0` is at `t=8`), so there is no
reason to place the recv ahead of the ready sends at all.

However, if I configure scheduler with `defer_pp_recv=True`, everything works fine.

```python
schedule = ScheduleInterleaved1F1B(stages, n_microbatches=mb, loss_fn=F.mse_loss, defer_pp_recv=True)
```

**The timeline of rank1**

<img width="941" height="94" alt="Image" src="https://github.com/user-attachments/assets/26419ef2-3e54-47d1-a829-a994b8163bd1" />

### Versions

```console
Collecting environment information...
PyTorch version: 2.14.0a0+git882cc6d
Is debug build: False
CUDA used to build PyTorch: 12.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.4 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04.1) 13.3.0
Clang version: Could not collect
CMake version: version 3.31.6
Libc version: glibc-2.39

Python version: 3.12.0 | packaged by Anaconda, Inc. | (main, Oct  2 2023, 17:29:18) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.15.0-89-generic-x86_64-with-glibc2.39
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: 


Nvidia driver version: 570.133.20
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.9.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv.so.9.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn.so.9.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_precompiled.so.9.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_runtime_compiled.so.9.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_graph.so.9.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_heuristic.so.9.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops.so.9.8.0
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False
Caching allocator config: {'PYTORCH_CUDA_ALLOC_CONF': 'expandable_segments:True'}

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      46 bits physical, 57 bits virtual
Byte Order:                         Little Endian
CPU(s):                             192
On-line CPU(s) list:                0
Off-line CPU(s) list:               1-191
Vendor ID:                          GenuineIntel
Model name:                         INTEL(R) XEON(R) PLATINUM 8558
CPU family:                         6
Model:                              207
Thread(s) per core:                 2
Core(s) per socket:                 48
Socket(s):                          2
Stepping:                           2
CPU(s) scaling MHz:                 99%
CPU max MHz:                        4000.0000
CPU min MHz:                        800.0000
BogoMIPS:                           4200.00
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 mo
nitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cat_l2 cdp_l3 invpcid_single intel_ppin cdp_l2 ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept 
vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect avx_vnni avx512_bf16 wbnoinvd dtherm i
da arat pln pts hwp hwp_act_window hwp_epp hwp_pkg_req avx512vbmi umip pku ospke waitpkg avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid bus_lock_detect cldemote movdiri movdir64b enqcmd fsrm md_clear serialize tsxldtrk pconfig arch_lbr amx_bf16 avx512_fp16 amx_tile amx_
int8 flush_l1d arch_capabilities
Virtualization:                     VT-x
L1d cache:                          4.5 MiB (96 instances)
L1i cache:                          3 MiB (96 instances)
L2 cache:                           192 MiB (96 instances)
L3 cache:                           520 MiB (2 instances)
NUMA node(s):                       2
NUMA node0 CPU(s):                  0-47,96-143
NUMA node1 CPU(s):                  48-95,144-191
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
CPU(s):                             192
On-line CPU(s) list:                0
Off-line CPU(s) list:               1-191
Vendor ID:                          GenuineIntel
Model name:                         INTEL(R) XEON(R) PLATINUM 8558
CPU family:                         6
Model:                              207
Thread(s) per core:                 2
Core(s) per socket:                 48
Socket(s):                          2
Stepping:                           2
CPU(s) scaling MHz:                 99%
CPU max MHz:                        4000.0000
CPU min MHz:                        800.0000
BogoMIPS:                           4200.00
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 mo
nitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cat_l2 cdp_l3 invpcid_single intel_ppin cdp_l2 ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept 
vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect avx_vnni avx512_bf16 wbnoinvd dtherm i
da arat pln pts hwp hwp_act_window hwp_epp hwp_pkg_req avx512vbmi umip pku ospke waitpkg avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid bus_lock_detect cldemote movdiri movdir64b enqcmd fsrm md_clear serialize tsxldtrk pconfig arch_lbr amx_bf16 avx512_fp16 amx_tile amx_
int8 flush_l1d arch_capabilities
Virtualization:                     VT-x
L1d cache:                          4.5 MiB (96 instances)
L1i cache:                          3 MiB (96 instances)
L2 cache:                           192 MiB (96 instances)
L3 cache:                           520 MiB (2 instances)
NUMA node(s):                       2
NUMA node0 CPU(s):                  0-47,96-143
NUMA node1 CPU(s):                  48-95,144-191
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Not affected
Vulnerability Retbleed:             Not affected
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] numpy==2.5.0
[pip3] torch==2.14.0a0+git882cc6d
[conda] numpy                     2.5.0                    pypi_0    pypi
[conda] torch                     2.14.0a0+git882cc6d          pypi_0    pypi

```

cc @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @aditvenk @weifengpy

**Comment by HAOCHENYE** [comment]
@H-Huang any thoughts on this?

**Comment by prabhaks** [comment]
Dug into the lowering and the mechanism checks out, and it isn't inherently AMD-specific.

In `_add_send_recv` the RECV is appended to the receiver's comm list at the iteration where the *sender's* compute action is processed, not where the receiver consumes it:

```python
# schedules.py ~L1536
comm_actions[stage_to_rank(recv.stage_index)].append(recv)
```

So during warmup a long-latency `RECV_F` gets ordered ahead of same-rank `SEND`s that are already runnable, which is the head-of-line bubble on the serialized p2p stream you're describing.

One caveat on "not AMD-specific": the ordering is suboptimal on any backend, but how much it *costs* on NCCL isn't guaranteed. Recvs are posted async and only waited at the consuming `FORWARD`, and `pipeline_per_direction_p2p` (plus the non-coalesced homogeneous batch path) can let SENDs proceed past a pending RECV. So the NVIDIA cost depends on stream/communicator layout, which is exactly the crux of the default question below.

On the fix: the deferral you already validated (`defer_pp_recv=True`) is `_defer_recv_ops` (~L1550-1641), which moves each RECV to just before its consuming compute op with rank-parity ordering to stay deadlock-safe. That landed in #178815 (resolving #172668) and is plumbed through all the multi-stage schedules including Interleaved1F1B. So for anyone hitting this today the flag is the answer, and it matches your result. The real open item is the *default*, and that's a policy call I don't want to guess at, especially since the `_defer_recv_ops` docstring (~L1558) frames ASAP recv posting as the deliberate default to overlap comm with compute, and #178815 shipped deferral as opt-in rather than flipping it.

A few directions, roughly in order of blast radius:

**A. Flip the default to defer (`defer_pp_recv=True`).** Trivial code, but `_defer_recv_ops` defers *all* recvs including steady state, giving up the ASAP overlap on the common NCCL path where a pending irecv doesn't block compute. High regression risk on NVIDIA, and it re-opens the default that #178815 review deliberately landed as opt-in. Not credible without cross-platform benchmarks.

**B. Warmup-scoped deferral.** Keep ASAP in steady state, only defer a RECV past same-rank ready SENDs during warmup. This matches the actual root cause here (the bubble is warmup-specific; steady-state overlap is still wanted). More involved in the lowering and deadlock-sensitive, since it has to preserve the rank-parity invariant `_defer_recv_ops` relies on or it risks hangs. Needs numbers on both stacks.

**C. Keep the default, document `defer_pp_recv=True`** as the mitigation for the interleaved warmup case, plus a lowering-time check/test that flags a long-latency RECV ordered ahead of ready SENDs so future regressions stay visible. Zero risk, but it doesn't address the ordering you're calling unreasonable.

@H-Huang before anyone writes code, which of these is in scope? If B is acceptable in principle I'm happy to prototype it and bring NVIDIA + AMD numbers, but I'd rather not send a default-flip PR that re-litigates the #178815 decision without your read on the tradeoff first.

**Comment by HAOCHENYE** [comment]
😢

**Comment by aditvenk** [comment]
cc @sanketpurandare

**Comment by HAOCHENYE** [comment]
😢

**Comment by sanketpurandare** [comment]
The options are purposefully a trade-off. If you can show via experimentation that one policy is always better than the other then we can make it as default. 

Also if you have a better general schedule we always provide an option of providing a schedule.csv that can be consumed by the Pipeline Schedule Runtime.
